### PR TITLE
deprecate write_tree_flag in favor of direct method arguments

### PR DESCRIPTION
**Pull request by joeljonsson**
Resolve some of the issues discussed in #160 and fixes current errors with examples (#162, but testing examples is still a good idea). Notable changes:

- remove all uses of the `write_tree_flag` and mark the set/get methods as deprecated
- `write_apr` now takes fourth argument:  
  ```bool write_apr(APR& apr,uint64_t t = 0,std::string channel_name = "t", bool write_tree = true);```  
- `read_apr` reads the APR and tries to read the tree. If the tree does not exist, ~~it initializes it from the APR structure~~ it is left uninitialized until required by other methods.
- for particle i/o and lazydata, dataset existence is checked and handled based on the `apr_or_tree` argument

The new behavior of `read_apr` has some performance implications. Reading the tree is quite fast (adds ~5-15% to overall read time), ~~but initializing the tree from the APR is significantly slower (makes overall read time up to 7 times longer). Perhaps the default behavior if the tree data is not on file should be to leave it uninitialized.~~ What do you think @cheesema?

### CODE REVIEW

**Comment by joeljonsson** [comment]
This resolves #164 
`read_apr` no longer initializes the tree. Instead, that us done in method that require it, or can be done manually.

@cheesema I think this is ready

File: src/data_structures/APR/APR.hpp [hunk 1]
```diff
@@ -25,11 +25,16 @@ class APR {
...
 protected:
 
-    //APR Tree function
-    void initialize_apr_tree_sparse();
-    void initialize_apr_tree_sparse_linear();
-    void initialize_apr_tree();
-    void initialize_linear_access(LinearAccess& aprAccess,APRIterator& it);
+    // initialize tree RandomAccess
+    void initialize_tree_random_sparse();
+    void initialize_tree_random_dense();    // appears to be broken #TODO: remove or fix
```
Comment (krzysg) [inline src/data_structures/APR/APR.hpp:30]: Could you please rather register issue for it?

**Comment by krzysg** [comment]
I do not have more specific comments. I do not really like how APR.hpp looks but** it is not connected to this specific changes but in general it ... screams for big refactoring which is not (and should not be) part of this PR.

File: src/io/APRFile.hpp [hunk 2]
```diff
@@ -121,7 +118,7 @@ class APRFile {
 
     //Basic Properties.
     uint64_t current_t=0;
-    bool with_tree_flag = true;
+    bool with_tree_flag = true; //usage is deprecated, keeping for backward compatibility
```
[the remark this answers is not in the document]
Comment (joeljonsson) [reply src/io/APRFile.hpp:121]: Good idea, I will merge this for now but definitely have to go through and remove unused and deprecated code before the next release

**Comment by joeljonsson** [comment]
Good points. I think large parts of the library need a major cleanup -- not just removing unused and deprecated code, but also restructuring and rethinking some classes. 

I will merge this for now though

File: src/data_structures/APR/APR.hpp (hunk 1 above)
Comment (joeljonsson) [reply src/data_structures/APR/APR.hpp:30]: done

**Pull request merged**